### ISSUE TITLE
CA verification failed

### ISSUE THREAD

**Issue by kvaps** [jet/kube-webhook-certgen#25]
Hi, please correct me if I'm wrong, I found that [nginx-ingress](https://github.com/kubernetes/ingress-nginx/tree/master/charts/ingress-nginx) is creating certificate like this:

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: nginx-ingress-ingress-nginx-admission-create
  namespace: ingress-nginx
spec:
  backoffLimit: 6
  completions: 1
  parallelism: 1
  template:
    metadata:
      creationTimestamp: null
      name: nginx-ingress-ingress-nginx-admission-create
    spec:
      containers:
      - args:
        - create
        - --host=nginx-ingress-ingress-nginx-controller-admission,nginx-ingress-ingress-nginx-controller-admission.$(POD_NAMESPACE).svc
        - --namespace=$(POD_NAMESPACE)
        - --secret-name=nginx-ingress-ingress-nginx-admission
        env:
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        image: docker.io/jettech/kube-webhook-certgen:v1.5.0
        imagePullPolicy: IfNotPresent
        name: create
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: OnFailure
      schedulerName: default-scheduler
      securityContext:
        runAsNonRoot: true
        runAsUser: 2000
      serviceAccount: nginx-ingress-ingress-nginx-admission
      serviceAccountName: nginx-ingress-ingress-nginx-admission
      terminationGracePeriodSeconds: 30
```

but issued certificate can't be verified:

```console
# kubectl -n ingress-nginx get secret nginx-ingress-ingress-nginx-admission -o go-template='{{ .data.ca | base64decode }}'  > /tmp/ca
# kubectl -n ingress-nginx get secret nginx-ingress-ingress-nginx-admission -o go-template='{{ .data.cert | base64decode }}'  > /tmp/cert
# openssl verify -CAfile /tmp/ca /tmp/cert

error 18 at 0 depth lookup: self signed certificate
error /tmp/cert: verification failed
```

v1.5.0 version has bit different output:
```console
# openssl verify -CAfile /tmp/ca /tmp/cert
O = nil
error 18 at 0 depth lookup: self signed certificate
error /tmp/cert: verification failed
```

**Comment by vsliouniaev** [comment]
I think that error indicates that the root cert didn't sign the leaf cert. 
`O = nil` is a change introduced in v1.4.0 to address (https://github.com/jet/kube-webhook-certgen/issues/18)

Appears to work fine on this version of openssl
```
$ openssl version
LibreSSL 2.8.3
```

The codebase contains a unit test to validate that generated certificates are valid according to go's validation mechanism. Kubernetes also appears to accept this certificate when used with its internal admission hooks.

**Comment by kvaps** [comment]
I have the latest **openssl** package from the archlinux repo:

```console
$ openssl version
OpenSSL 1.1.1h  22 Sep 2020
```
However it also has **libressl** package:
```console
$ libressl-openssl version
LibreSSL 3.2.2
```

and they behave differently:
```console
$ openssl verify -CAfile /tmp/ca /tmp/cert
O = nil
error 18 at 0 depth lookup: self signed certificate
error /tmp/cert: verification failed

$ libressl-openssl verify -CAfile /tmp/ca /tmp/cert
/tmp/cert: OK
```

To be honest, I'm bit confused. Why is libressl reports that certificate is OK, while openssl responds that certificate is self-signed?

**Comment by vsliouniaev** [comment]
I've managed to narrow this down to `OpenSSL 1.1.1h` not working but `OpenSSL 1.1.1g` working. [Looking at the release notes](https://www.openssl.org/news/openssl-1.1.1-notes.html):
1. Disallow explicit curve parameters in verifications chains when X509_V_FLAG_X509_STRICT is used (https://github.com/openssl/openssl/issues/12139)
2. Enable 'MinProtocol' and 'MaxProtocol' to configure both TLS and DTLS contexts
3. Oracle Developer Studio will start reporting deprecation warnings

The only one I can think of that would affect it is 1. However, the certificates do not use a custom curve and set the relevant public key info. And anyway, the error should be entirely different if a custom curve was used
```console
$ openssl x509 -in /tmp/ca -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            a9:e9:41:d6:55:af:38:16:c8:5b:65:d7:ba:e6:94:ef
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: O = nil
        Validity
            Not Before: Nov  6 00:28:45 2020 GMT
            Not After : Oct 13 00:28:45 2120 GMT
        Subject: O = nil
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:d7:08:5f:ac:36:15:7d:fe:c7:18:db:eb:3e:4d:
                    c6:e6:20:0f:a2:cc:ca:fe:99:53:84:27:22:5a:1a:
                    22:d2:35:ac:50:3b:e6:e4:ba:8c:74:6d:3b:28:36:
                    f7:cf:5b:9f:ea:74:81:11:ea:04:f8:01:06:07:9c:
                    1f:ce:9d:01:3b
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Certificate Sign
            X509v3 Extended Key Usage:
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier:
                0B:28:AD:8C:80:96:5C:DA:8C:98:BC:57:E2:3E:29:95:51:2E:36:4D
    Signature Algorithm: ecdsa-with-SHA256
         30:46:02:21:00:e1:0b:0a:5c:c9:85:48:ea:eb:c7:79:c9:51:
         83:de:94:11:cd:1e:10:65:e8:11:11:2d:b0:42:b3:44:3d:b5:
         f3:02:21:00:ee:24:39:46:e7:20:4b:3c:b5:c9:78:65:ef:2e:
         7c:97:29:d6:8a:e8:22:5d:c6:1e:15:f4:cd:93:83:0e:23:6d
```
```console
$ openssl x509 -in /tmp/cert -text -noout
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            ca:bc:44:89:b1:26:bf:36:23:dc:ce:d4:52:15:91:87
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: O = nil
        Validity
            Not Before: Nov  6 00:28:45 2020 GMT
            Not After : Oct 13 00:28:45 2120 GMT
        Subject: O = nil
        Subject Public Key Info:
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:4d:02:f6:1e:5b:f1:e1:96:35:f8:e1:e5:ab:2f:
                    b8:0d:3f:da:08:77:db:c0:c7:c5:0e:0d:32:24:db:
                    a2:6d:3f:49:f7:91:8c:e9:e0:6a:dc:8b:ed:17:0d:
                    27:b5:73:cf:24:41:fb:0f:47:63:2a:6e:c2:ba:34:
                    dd:0f:8c:33:18
                ASN1 OID: prime256v1
                NIST CURVE: P-256
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Alternative Name:
                DNS:localhost
    Signature Algorithm: ecdsa-with-SHA256
         30:45:02:20:5a:6c:63:b2:3f:68:b5:b8:13:d2:57:8f:46:ca:
         f9:14:3e:f0:c1:0a:b3:c2:e4:c8:f1:3b:44:57:61:0b:7f:6d:
         02:21:00:d5:9a:27:b8:60:f2:f3:d0:ad:03:9e:bd:cc:36:73:
         43:16:3d:a5:2e:1f:4c:63:46:a6:c4:9a:48:40:0e:de:87
```

**Comment by vsliouniaev** [comment]
Looks like the issue appears to happen when the Subject and Issuer match in the client cert. Could you please check if v1.5.1 works for you?

**Comment by vsliouniaev** [comment]
Fixed in `v1.5.1`
